client.go

```diff
@@ -0,0 +1,593 @@
+package milter
+
+import (
+	"bytes"
+	"encoding/binary"
+	"fmt"
+	"io"
+	"net"
+	"strconv"
+	"time"
+
+	"github.com/emersion/go-message/textproto"
+)
+
+// Client is a wrapper for managing milter connections.
+//
+// Currently it just creates new connections using provided Dialer.
+type Client struct {
+	// Dialer is used to establish new connections to the milter.
+	// Set to empty net.Dialer{} by NewClient.
+	Dialer interface {
+		Dial(network string, addr string) (net.Conn, error)
+	}
+
+	ReadTimeout  time.Duration
+	WriteTimeout time.Duration
+
+	Network string
+	Address string
+}
+
+// NewClient creates a new Client instance populating fields with default
+// values.
+//
+// Dialer is net.Dialer with 10 second timeout, Read and Write timeouts are set
+// to 10 seconds too.
+func NewClient(network, address string) *Client {
+	return &Client{
+		Dialer: &net.Dialer{
+			Timeout: 10 * time.Second,
+		},
+		ReadTimeout:  10 * time.Second,
+		WriteTimeout: 10 * time.Second,
+		Network:      network,
+		Address:      address,
+	}
+}
+
+func (c *Client) Session(actionMask OptAction, protoMask OptProtocol) (*ClientSession, error) {
+	s := &ClientSession{
+		readTimeout:  c.ReadTimeout,
+		writeTimeout: c.WriteTimeout,
+	}
+
+	// TODO(foxcpp): Connection pooling.
```

Comment:
As-is, the client network/addr can be changed on-the-fly by the library user. Will this prevent connection pooling?

Reply:
This can make Dialer changes no-op after first session. This should be documented.

Reply:
I wonder if having `Dial()` callback instead of Dialer + Network + Address is better. WDYT?

Reply:
I wonder if we should export these `Client` fields tbh. Maybe having just this would be enough:

```go
func Dial(network, addr string) (*Client, error)
func DialWithDialer(dialer Dialer, network, addr string) (*Client, error)
```

Reply:
Yeah, given placeholders for connection pooling. That will be much better.

Reply:
Eh, that still doesn't allow the user to customize timeouts. What about:

```
func Dial(network, addr string, options *Options) (*Client, error)
func DialWithDialer(dialer Dialer, network, addr string, options *Options) (*Client, error)
```

Reply:
At which point, the dialer may just be part of options too.

Reply:
What's about keeping it like go-smtp? With exported *Timeout fields and obvious caveat these should not be touched past initial configuration.

Reply:
Works for me

Reply:
Actually, on second thought I think Options structure is better. Notably because arguments of Session function should be actually moved into it.

Reply:
Oh yeah that makes sense. We'd need to do per-session-args pools otherwise, which would be cumbersome.

Reply:
What do you think about 119a7a8?